Improve bookshelf display - proper display when detail page is navigated to directly (BL-3863)

andrew-polk opened the pull request.

JohnThomson commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/modules/common/services.js, line 641 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KTh_Ode7PRgwxcNKw1a:-KTh_Ode7PRgwxcNKw1b:b-anco7y) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L641)):*
> ```JavaScript
> 
>         var cachedBookshelves;
>         this.getBookshelfDisplayName = function(shelfKey) {
> ```

I think it would be helpful somewhere around here to explain the two situations being dealt with (i.e., in what circumstances do we get the bookshelves from $rootScope, and when do we have to retrieve and cache them. 
Why do we need the $rootScope option any way? Could we just abolish $rootScope.bookshelves in favor of a service that returns the available bookshelves? There may be a good reason (e.g., performance, or a lot of code to change), but I hate to see two ways of getting at the same information if we don't have to.
If we do have to have two ways, it would still be good I think to create a separate method (even if it only has one client so far) to encapsulate the idea of "get the global collection of all existing bookshelves." Hmmm...I guess that's what getBookshelves does. Would it make sense to put the smarts about caching it and possibly getting it from rootScope there? Or is that method used to populate rootScope in the other path?
Not a showstopper, but worth a thought or two...

---

*[src/app/modules/common/services.js, line 642 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KThZ_QyS__YTHgA_P4y:-KThZ_QyS__YTHgA_P4z:bxt1u60) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L642)):*
> ```JavaScript
>         var cachedBookshelves;
>         this.getBookshelfDisplayName = function(shelfKey) {
>             if (cachedBookshelves) {                
> ```

Seems to be trailing white space here...good thing to avoid

---

*[src/app/modules/common/services.js, line 654 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KTh_1Z5_0eDxXWBQ6tT:-KTh_1Z5_0eDxXWBQ6tU:b8afww7) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L654)):*
> ```JavaScript
>                     return getBookshelfDisplayNameInternal(shelfKey, bookshelves);
>                 });
>             }            
> ```

trailing space again

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Review status: 0 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/modules/common/services.js, line 641 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KTh_Ode7PRgwxcNKw1a:-KTj-tElwKgFFMJ6alNv:bd17svb) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L641)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> I think it would be helpful somewhere around here to explain the two situations being dealt with (i.e., in what circumstances do we get the bookshelves from $rootScope, and when do we have to retrieve and cache them. 
> Why do we need the $rootScope option any way? Could we just abolish $rootScope.bookshelves in favor of a service that returns the available bookshelves? There may be a good reason (e.g., performance, or a lot of code to change), but I hate to see two ways of getting at the same information if we don't have to.
> If we do have to have two ways, it would still be good I think to create a separate method (even if it only has one client so far) to encapsulate the idea of "get the global collection of all existing bookshelves." Hmmm...I guess that's what getBookshelves does. Would it make sense to put the smarts about caching it and possibly getting it from rootScope there? Or is that method used to populate rootScope in the other path?
> Not a showstopper, but worth a thought or two...
</details>
Some good thoughts. I'll see what I can do.

---

*[src/app/modules/common/services.js, line 642 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KThZ_QyS__YTHgA_P4y:-KTj-lEhbQp2SJCNlaLn:bw58ohr) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L642)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> Seems to be trailing white space here...good thing to avoid
</details>
Hm. Looks like this repository isn't set up to check for trailing space. Fixed.

---

*[src/app/modules/common/services.js, line 654 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KTh_1Z5_0eDxXWBQ6tT:-KTj-sBmcBaFpE7dRk5r:bw58ohr) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L654)):*
<details><summary><i>Previously, JohnThomson (John Thomson) wrote…</i></summary>
> trailing space again
</details>
Hm. Looks like this repository isn't set up to check for trailing space. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252)*
<!-- Sent from Reviewable.io -->

andrew-polk commented:
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/modules/common/services.js, line 641 at r1](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KTh_Ode7PRgwxcNKw1a:-KTjBG_gaGXJ6eVn3Ycn:bgk7mm5) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/d6b19643e32cde6478a21a6a91e070670b409c27/src/app/modules/common/services.js#L641)):*
<details><summary><i>Previously, andrew-polk wrote…</i></summary>
> Some good thoughts. I'll see what I can do.
</details>
Improved.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252)*
<!-- Sent from Reviewable.io -->

hatton commented:
John, can you take another look at this? Thanks.

JohnThomson commented:
Much improved. I had a couple more ideas, but if they seem too much trouble, could go ahead as is.

---

Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/app/modules/app/app.js, line 181 at r2](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KUU4iNqsILC6SjXgknq:-KUU4iNqsILC6SjXgknr:b-s635yk) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/4ab2f5c9eb5d84597f74935cf97b47011b707194/src/app/modules/app/app.js#L181)):*
> ```JavaScript
>             });
> 
>             bookService.getBookshelves().then(function(bookshelves) {
> ```

Is there still any reason for bookshelves to be a parameter of this function? I found it quite confusing.
Conversely, I think the way you resolve the defer means that the parameter of the 'then' function WILL be the localized bookshelves, in which case, the following line and comment could be removed.

---

*[src/app/modules/common/services.js, line 231 at r2](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KUU7S_oB3t2XGfRroK4:-KUU7S_oB3t2XGfRroK5:b8nlp) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/4ab2f5c9eb5d84597f74935cf97b47011b707194/src/app/modules/common/services.js#L231)):*
> ```JavaScript
>         };
> 
>         this.getBookshelves = function() {
> ```

If you made this function check to see whether $rootscope already had cachedAndLocalizedBookshelves, the existence of that cache could be a secret of this function. Just return an already-resolved promise if you already have the data. (Or is it somewhere used when you want to force reloading the cache?)

---

*[src/app/modules/common/services.js, line 651 at r2](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-KUU5DSfYcgnKpAE2ZoS:-KUU5DSfYcgnKpAE2ZoT:b-cu6t63) ([raw file](https://github.com/bloombooks/bloomlibrary/blob/4ab2f5c9eb5d84597f74935cf97b47011b707194/src/app/modules/common/services.js#L651)):*
> ```JavaScript
>             } else {
>                 // The call to getBookshelves() populates $rootScope.cachedAndLocalizedBookshelves
>                 this.getBookshelves().then(function(bookshelves) {
> ```

Again, it appears this function does not need an argument..or probably better, use the argument

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/bloombooks/bloomlibrary/252#-:-KUU83rDvVWOuMJlW65T:b-vpu6pm)*
<!-- Sent from Reviewable.io -->
